src/jdk.jdwp.agent/share/native/libjdwp/commonRef.c

```diff
@@ -176,20 +185,30 @@ strengthenNode(JNIEnv *env, RefNode *node)
         if (strongRef != NULL) {
             JNI_FUNC_PTR(env,DeleteWeakGlobalRef)(env, node->ref);
             node->ref         = strongRef;
-            node->strongCount = 1;
         }
-        return strongRef;
```

Comment:
Just to be clear about the removal of this line, at this point if `strongRef==NULL` then `node->ref==NULL` (otherwise we would have taken the EXIT_ERROR path) - right?

Reply:
I believe `node->ref` would be a weak reference to a NULL object, and not itself be NULL. That is why the !isSameObject() call is used rather than just comparing to NULL. It's an error if the global ref allocation fails and `node->ref` does not reference a NULL (collected) object.

However, there does appear to be a bug here. If the allocation failed because the object was collected, we want to return NULL in that case, not node->ref, which is not NULL but references a NULL object. Also, I it's a bug to set isPinAll or isCommonPin when the allocation fails for this reason.